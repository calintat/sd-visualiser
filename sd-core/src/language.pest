variable = @{ !("bind" | "in") ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "_")* }
bind_clause = { "bind" ~ variable ~ "=" ~ term ~ "in" }
expr = { bind_clause* ~ value }
term = { thunk | (active_op ~ "(" ~ args ~ ")") | value }
value = { (passive_op ~ "(" ~ args ~ ")") | variable }
active_op = { "plus" | "minus" | "times" | "eq" | "if" | "app" | "lambda" | "rec" }
passive_op = { number }
number = _{ ASCII_DIGIT+ }
args = _{ value ~ ("," ~ value)* | "" }
program = _{ SOI ~ expr ~ EOI }
WHITESPACE = _{ " " | "\t" | "\n" }
thunk = { variable* ~ "." ~ expr }