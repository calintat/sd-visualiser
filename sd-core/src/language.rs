pub mod chil;
pub mod spartan;

pub(super) mod visitor;
