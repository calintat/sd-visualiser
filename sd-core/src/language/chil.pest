program = _{ SOI ~ expr ~ EOI }

keyword = { "def" | "output" | "thunk" }

expr = { bind_clause* ~ "output" ~ value }
bind_clause = { "def" ~ variable_def ~ "=" ~ value }

thunk = { "thunk" ~ addr ~ "=" ~ "{" ~ (variable_def ~ ("," ~ variable_def)*)? ~ "=>" ~ expr ~ "}" }

value = { variable | op ~ ("(" ~ (thunk+ | values ~ (";" ~ thunk+)?) ~ ")")? }
values = _{ value ~ ("," ~ value)* }

op = ${ !keyword ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "+" | "-" | "*" | "=")* ~ ("/" ~ (number | string | ty ~ "/" ~ ty))? }
number = @{ ASCII_DIGIT+ ~ ("." ~ ASCII_DIGIT+)? }
string = { "\"" ~ (!"\"" ~ ANY)* ~ "\"" }

ty = { generic_ty | base_ty | function_ty | tuple_ty }
tys = _{ ty ~ ("," ~ ty)* }
base_ty = @{ 'A'..'Z' ~ (ASCII_ALPHANUMERIC | "-")* }
generic_ty = { base_ty ~ "<" ~ tys ~ ">" }
tuple_ty = { "(" ~ tys? ~ ")" }
function_ty = { tuple_ty ~ "->" ~ ty }

variable = { addr | identifier ~ "(" ~ "id" ~ ":" ~ addr ~ ")" }
variable_def = { variable ~ ":" ~ ty | variable }

COMMENT = _{ "#" ~ (!"\n" ~ ANY)* }
WHITESPACE = _{ " " | "\t" | "\n" }

addr = @{ ("@" | "%") ~ ASCII_DIGIT+ }
identifier = @{ !keyword ~ "_"* ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "_")* }