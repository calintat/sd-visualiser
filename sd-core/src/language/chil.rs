#![allow(clippy::clone_on_copy)]

use std::fmt::{Display, Write};

use from_pest::{ConversionError, FromPest, Void};
use pest::iterators::Pairs;
use pest_ast::FromPest;
use pest_derive::Parser;

use super::{span_into_str, AsVar};

pub struct Chil;

impl super::Language for Chil {
    type Op = Op;
    type Var = Variable;
    type Addr = Addr;
    type VarDef = VariableDef;

    type Rule = Rule;

    fn expr_rule() -> Self::Rule {
        Rule::expr
    }
    fn bind_rule() -> Self::Rule {
        Rule::bind
    }
    fn value_rule() -> Self::Rule {
        Rule::value
    }
    fn thunk_rule() -> Self::Rule {
        Rule::thunk
    }
}

pub type Expr = super::Expr<Chil>;
pub type Bind = super::Bind<Chil>;
pub type Value = super::Value<Chil>;
pub type Thunk = super::Thunk<Chil>;

#[derive(Parser)]
#[grammar = "language/chil.pest"]
pub struct ChilParser;

fn parse_addr_first(input: &str) -> char {
    input.chars().next().unwrap()
}

fn parse_addr_second(input: &str) -> usize {
    input[1..].parse().unwrap()
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub struct Op(pub String);

impl Display for Op {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.0.as_str() {
            "+" | "throwing+" => f.write_char('+'),
            "-" | "throwing-" => f.write_char('-'),
            "*" | "throwing*" => f.write_char('*'),
            "/" | "throwing/" => f.write_char('/'),
            "%" | "throwing%" => f.write_char('%'),
            "&&" | "throwing&&" => f.write_str("&&"),
            "||" | "throwing||" => f.write_str("||"),
            "!" | "throwing!" => f.write_char('!'),
            "==" | "throwing==" => f.write_char('='),
            "!=" | "throwing!=" => f.write_char('≠'),
            "<" | "throwing<" => f.write_char('<'),
            "<=" | "throwing<=" => f.write_char('≤'),
            ">" | "throwing>" => f.write_char('>'),
            ">=" | "throwing>=" => f.write_char('≥'),
            "func" => f.write_char('λ'),
            "seq" => f.write_char(';'),
            "unit" => f.write_str("()"),
            str => {
                if str.starts_with("apply/") {
                    return f.write_char('@');
                }
                if str.starts_with("tuple/") {
                    return f.write_char('⊗');
                }
                if let Some(rest) = None
                    .or_else(|| str.strip_prefix("bool/"))
                    .or_else(|| str.strip_prefix("int64/"))
                    .or_else(|| str.strip_prefix("float64/"))
                    .or_else(|| str.strip_prefix("string/"))
                {
                    return f.write_str(rest);
                }
                f.write_str(str)
            }
        }
    }
}

impl<'pest> FromPest<'pest> for Op {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::op {
            return Err(ConversionError::NoMatch);
        }
        *pest = clone;
        Ok(Self(pair.as_str().to_owned()))
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::variable))]
pub struct Variable {
    pub name: Option<Identifier>,
    pub addr: Addr,
}

impl Display for Variable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.name {
            None => write!(f, "{}", self.addr),
            Some(name) => write!(f, "{}(id: {})", name, self.addr),
        }
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::addr))]
pub struct Addr(
    #[pest_ast(outer(with(span_into_str), with(parse_addr_first)))] pub char,
    #[pest_ast(outer(with(span_into_str), with(parse_addr_second)))] pub usize,
);

impl Display for Addr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", self.0, self.1)
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub struct Identifier(pub String);

impl Display for Identifier {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl<'pest> FromPest<'pest> for Identifier {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::identifier {
            return Err(ConversionError::NoMatch);
        }
        *pest = clone;
        Ok(Self(pair.as_str().to_owned()))
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::variable_def))]
pub struct VariableDef {
    pub var: Variable,
    pub r#type: Option<Type>,
}

impl AsVar<Variable> for VariableDef {
    fn as_var(&self) -> &Variable {
        &self.var
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::ty))]
pub enum Type {
    Base(BaseType),
    Generic(GenericType),
    Tuple(TupleType),
    Function(FunctionType),
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::base_ty))]
pub struct BaseType(#[pest_ast(outer(with(span_into_str), with(str::to_string)))] pub String);

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::generic_ty))]
pub struct GenericType {
    pub base: BaseType,
    pub params: Vec<Type>,
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::tuple_ty))]
pub struct TupleType {
    pub types: Vec<Type>,
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::function_ty))]
pub struct FunctionType {
    pub domain: TupleType,
    pub codomain: Box<Type>,
}

// Conversion to spartan

#[cfg(test)]
impl From<Op> for super::spartan::Op {
    fn from(op: Op) -> Self {
        match op.0.as_str() {
            "+" | "throwing+" => return Self::Plus,
            "-" | "throwing-" => return Self::Minus,
            "*" | "throwing*" => return Self::Times,
            "/" | "throwing/" => return Self::Div,
            "%" | "throwing%" => return Self::Rem,
            "&&" | "throwing&&" => return Self::And,
            "||" | "throwing||" => return Self::Or,
            "!" | "throwing!" => return Self::Not,
            "==" | "throwing==" => return Self::Eq,
            "!=" | "throwing!=" => return Self::Neq,
            "<" | "throwing<" => return Self::Lt,
            "<=" | "throwing<=" => return Self::Leq,
            ">" | "throwing>" => return Self::Gt,
            ">=" | "throwing>=" => return Self::Geq,
            "func" => return Self::Lambda,
            "atom" => return Self::Atom,
            "deref" => return Self::Deref,
            "asg" => return Self::Assign,
            "bool/true" => return Self::Bool(true),
            "bool/false" => return Self::Bool(false),
            str => {
                if str.starts_with("apply/") {
                    return Self::App;
                }
                if let Some(rest) = str.strip_prefix("int64/") {
                    return Self::Number(rest.parse().unwrap_or_default());
                }
                Self::Plus // dummy placeholder
            }
        }
    }
}

#[cfg(test)]
impl From<Variable> for super::spartan::Variable {
    fn from(var: Variable) -> Self {
        Self(format!("var_{}", var.addr.1))
    }
}

#[cfg(test)]
impl From<Addr> for super::spartan::Addr {
    fn from(_addr: Addr) -> Self {
        Self
    }
}

#[cfg(test)]
impl From<VariableDef> for super::spartan::Variable {
    fn from(def: VariableDef) -> Self {
        def.var.into()
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::path::Path;

    use dir_test::{dir_test, Fixture};
    use from_pest::FromPest;
    use pest::Parser;

    use super::{ChilParser, Expr, Rule};

    pub fn parse_chil(raw_path: &str) -> (&str, Expr) {
        let path = Path::new(raw_path);
        let program = std::fs::read_to_string(path).unwrap();
        let mut pairs = ChilParser::parse(Rule::program, &program).unwrap_or_else(|err| {
            panic!(
                "could not parse program {:?}\n{err:?}",
                path.file_stem().unwrap()
            )
        });
        let name = path.file_stem().unwrap().to_str().unwrap();
        let expr = Expr::from_pest(&mut pairs).unwrap();
        (name, expr)
    }

    #[allow(clippy::needless_pass_by_value)]
    #[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "**/*.chil", loader: crate::language::chil::tests::parse_chil, postfix: "check_parse")]
    fn check_parse(fixture: Fixture<(&str, Expr)>) {
        let (_name, _expr) = fixture.content();
    }
}
