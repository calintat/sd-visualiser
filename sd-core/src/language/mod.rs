use std::{
    fmt::{Debug, Display},
    hash::Hash,
};

use derivative::Derivative;

use crate::{common::Matchable, prettyprinter::PrettyPrint};

pub mod chil;
pub mod spartan;

pub(crate) fn span_into_str(span: pest::Span) -> &str {
    span.as_str()
}

pub trait GetVar<V> {
    fn var(&self) -> &V;
    fn into_var(self) -> V;
}

impl<V> GetVar<V> for V {
    fn var(&self) -> &V {
        self
    }

    fn into_var(self) -> V {
        self
    }
}

pub trait Fresh {
    fn fresh(number: usize) -> Self;
}

/// `Display` should give the symbolic representation (e.g. "+").
/// `PrettyPrint` should give the textual representation (e.g. "plus").
pub trait Syntax:
    Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint
{
}
impl<T: Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint> Syntax for T {}

pub struct BlockWithArgs<T: Language + ?Sized> {
    pub block: T::BlockAddr,
    pub args: Vec<Value<T>>,
}

pub enum CF<T: Language + ?Sized> {
    Return,
    Brs(Vec<BlockWithArgs<T>>),
}

pub trait ControlFlow<T: Language + ?Sized> {
    fn get_cf(&self) -> Option<CF<T>>;

    #[must_use]
    fn get_apply() -> Option<T::Op> {
        None
    }
}

pub trait Language {
    type Op: Syntax + ControlFlow<Self>;
    type Var: Syntax + Fresh;
    type Addr: Syntax;
    type BlockAddr: Syntax;
    type VarDef: Syntax + GetVar<Self::Var>;
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Expr<T: Language + ?Sized> {
    pub binds: Vec<Bind<T>>,
    pub values: Vec<Value<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Bind<T: Language + ?Sized> {
    pub defs: Vec<T::VarDef>,
    pub value: Value<T>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub enum Value<T: Language + ?Sized> {
    Variable(T::Var),
    Thunk(Thunk<T>),
    Op { op: T::Op, args: Vec<Value<T>> },
}

impl<T: Language> ControlFlow<T> for Value<T> {
    fn get_cf(&self) -> Option<CF<T>> {
        match self {
            Value::Op { op, .. } => op.get_cf(),
            _ => None,
        }
    }
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Thunk<T: Language + ?Sized> {
    pub addr: T::Addr,
    pub args: Vec<T::VarDef>,
    pub body: Expr<T>,
    pub blocks: Vec<Block<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Block<T: Language + ?Sized> {
    pub addr: T::BlockAddr,
    pub args: Vec<T::VarDef>,
    pub expr: Expr<T>,
}

// Conversions between languages

impl<T: Language> Expr<T> {
    pub fn into<U: Language>(self) -> Expr<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Expr {
            binds: self.binds.into_iter().map(Bind::into).collect(),
            values: self.values.into_iter().map(Value::into).collect(),
        }
    }
}

impl<T: Language> Bind<T> {
    pub fn into<U: Language>(self) -> Bind<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Bind {
            defs: self.defs.into_iter().map(Into::into).collect(),
            value: self.value.into(),
        }
    }
}

impl<T: Language> Value<T> {
    pub fn into<U: Language>(self) -> Value<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        match self {
            Self::Variable(var) => Value::Variable(var.into()),
            Self::Thunk(thunk) => Value::Thunk(thunk.into()),
            Self::Op { op, args } => Value::Op {
                op: op.into(),
                args: args.into_iter().map(Value::into).collect(),
            },
        }
    }
}

impl<T: Language> Thunk<T> {
    pub fn into<U: Language>(self) -> Thunk<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Thunk {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            body: self.body.into(),
            blocks: self.blocks.into_iter().map(Block::into).collect(),
        }
    }
}

impl<T: Language> Block<T> {
    pub fn into<U: Language>(self) -> Block<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Block {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            expr: self.expr.into(),
        }
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::{ffi::OsStr, path::Path};

    use super::{
        chil::tests::parse_chil,
        spartan::{self, tests::parse_sd},
    };

    pub fn parse(raw_path: &str) -> (&str, &str, spartan::Expr) {
        let path = Path::new(raw_path);
        match path.extension() {
            Some(ext) if ext == OsStr::new("sd") => {
                let (name, expr) = parse_sd(raw_path);
                ("sd", name, expr)
            }
            Some(ext) if ext == OsStr::new("chil") => {
                let (name, expr) = parse_chil(raw_path);
                ("chil", name, expr.into())
            }
            _ => unreachable!(),
        }
    }
}
