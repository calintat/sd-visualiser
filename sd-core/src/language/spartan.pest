program = _{ SOI ~ expr ~ EOI }

keyword = { "bind" | "in" }

expr = { bind* ~ value }
bind = { "bind" ~ variable ~ "=" ~ value ~ "in" }

thunk = { variable* ~ "." ~ expr }

value = { variable | op ~ ("(" ~ (thunks | values ~ (";" ~ thunks)?) ~ ")")? }
thunks = _{ thunk ~ ("," ~ thunk)* }
values = _{ value ~ ("," ~ value)* }

op = { "plus" | "minus" | "times" | "div" | "rem" | "and" | "or" | "not" | "eq" | "neq" | "lt" | "leq" | "gt" | "geq" | "app" | "lambda" | "if" | "true" | "false" | number | string }
number = @{ ASCII_DIGIT+ ~ ("." ~ ASCII_DIGIT+)? }
string = { "\"" ~ (!"\"" ~ ANY)* ~ "\"" }

variable = @{ !(keyword | op) ~ "_"* ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "_")* }

COMMENT = _{ "#" ~ (!"\n" ~ ANY)* }
WHITESPACE = _{ " " | "\t" | "\n" }