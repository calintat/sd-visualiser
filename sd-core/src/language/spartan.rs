#![allow(clippy::clone_on_copy)]

use std::fmt::{Display, Write};

use from_pest::Void;
use ordered_float::NotNaN;
use pest_ast::FromPest;
use pest_derive::Parser;
#[cfg(test)]
use serde::Serialize;

use super::span_into_str;

#[derive(Parser)]
#[grammar = "language/spartan.pest"]
pub struct SpartanParser;

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::expr))]
pub struct Expr {
    pub binds: Vec<BindClause>,
    pub value: Value,
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::bind_clause))]
pub struct BindClause {
    pub var: Variable,
    pub value: Value,
}

#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::variable))]
pub struct Variable(#[pest_ast(outer(with(span_into_str), with(str::to_string)))] pub String);

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum Op {
    Plus,
    Minus,
    Times,
    Div,
    Rem,
    And,
    Or,
    Not,
    Eq,
    Neq,
    Lt,
    Leq,
    Gt,
    Geq,
    App,
    Lambda,
    Unit,
    Seq,
    Atom,
    Deref,
    Assign,
    Bool(bool),
    Number(NotNaN<f64>),
    String(String),     // string literal
    Identifier(String), // any other identifier
}

impl From<&str> for Op {
    fn from(str: &str) -> Self {
        match str {
            "plus" => Self::Plus,
            "minus" => Self::Minus,
            "times" => Self::Times,
            "div" => Self::Div,
            "rem" => Self::Rem,
            "and" => Self::And,
            "or" => Self::Or,
            "not" => Self::Not,
            "eq" => Self::Eq,
            "neq" => Self::Neq,
            "lt" => Self::Lt,
            "leq" => Self::Leq,
            "gt" => Self::Gt,
            "geq" => Self::Geq,
            "app" => Self::App,
            "lambda" => Self::Lambda,
            "unit" => Self::Unit,
            "seq" => Self::Seq,
            "atom" => Self::Atom,
            "deref" => Self::Deref,
            "assign" => Self::Assign,
            "true" => Self::Bool(true),
            "false" => Self::Bool(false),
            _ => {
                if let Ok(n) = str::parse::<f64>(str) {
                    return Self::Number(NotNaN::new(n).unwrap());
                }
                if str.starts_with('"') && str.ends_with('"') {
                    return Self::String(str[1..str.len() - 1].to_string());
                }
                Self::Identifier(str.to_owned())
            }
        }
    }
}

impl<'pest> from_pest::FromPest<'pest> for Op {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut pest::iterators::Pairs<'pest, Self::Rule>,
    ) -> Result<Self, from_pest::ConversionError<Self::FatalError>> {
        pest.next()
            .map(|pair| pair.as_str().into())
            .ok_or(from_pest::ConversionError::NoMatch)
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::value))]
pub enum Value {
    Variable(Variable),
    Op {
        op: Op,
        vs: Vec<Value>,
        ds: Vec<Thunk>,
    },
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::thunk))]
pub struct Thunk {
    pub args: Vec<Variable>,
    pub body: Expr,
}

impl Display for Op {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Plus => f.write_char('+'),
            Self::Minus => f.write_char('-'),
            Self::Times => f.write_char('×'),
            Self::Div => f.write_char('/'),
            Self::Rem => f.write_char('%'),
            Self::And => f.write_char('∧'),
            Self::Or => f.write_char('∨'),
            Self::Not => f.write_char('¬'),
            Self::Eq => f.write_char('='),
            Self::Neq => f.write_char('≠'),
            Self::Lt => f.write_char('<'),
            Self::Leq => f.write_char('≤'),
            Self::Gt => f.write_char('>'),
            Self::Geq => f.write_char('≥'),
            Self::App => f.write_char('@'),
            Self::Lambda => f.write_char('λ'),
            Self::Unit => f.write_str("()"),
            Self::Seq => f.write_char(';'),
            Self::Atom => f.write_char('&'),
            Self::Deref => f.write_char('!'),
            Self::Assign => f.write_str(":="),
            Self::Bool(b) => f.write_str(&b.to_string()),
            Self::Number(n) => f.write_str(&n.to_string()),
            Self::String(_) => f.write_str("str"),
            Self::Identifier(op) => f.write_str(op),
        }
    }
}

impl Display for Variable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl From<&str> for Variable {
    fn from(value: &str) -> Self {
        Variable(value.to_owned())
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::path::Path;

    use dir_test::{dir_test, Fixture};
    use from_pest::FromPest;
    use pest::Parser;

    use super::{Expr, Rule, SpartanParser};

    pub fn parse_sd(raw_path: &str) -> (&str, Expr) {
        let path = Path::new(raw_path);
        let program = std::fs::read_to_string(path).unwrap();
        let mut pairs = SpartanParser::parse(Rule::program, &program).unwrap_or_else(|err| {
            panic!(
                "could not parse program {:?}\n{err:?}",
                path.file_stem().unwrap()
            )
        });
        let name = path.file_stem().unwrap().to_str().unwrap();
        let expr = Expr::from_pest(&mut pairs).unwrap();
        (name, expr)
    }

    #[allow(clippy::needless_pass_by_value)]
    #[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "**/*.sd", loader: crate::language::spartan::tests::parse_sd, postfix: "check_parse")]
    fn check_parse(fixture: Fixture<(&str, Expr)>) {
        let (_name, _expr) = fixture.content();
    }
}
