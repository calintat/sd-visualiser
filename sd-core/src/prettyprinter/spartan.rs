use pretty::RcDoc;

use super::PrettyPrint;
use crate::language::spartan::{Bind, Expr, Op, Thunk, Value, Variable};

impl PrettyPrint for Expr {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::concat(self.binds.iter().map(PrettyPrint::to_doc)).append(RcDoc::intersperse(
            self.values.iter().map(PrettyPrint::to_doc),
            RcDoc::text(",").append(RcDoc::space()),
        ))
    }
}

impl PrettyPrint for Bind {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text("bind")
            .append(RcDoc::space())
            .append(self.def.to_doc())
            .append(RcDoc::space())
            .append(RcDoc::text("="))
            .append(RcDoc::space())
            .append(self.value.to_doc())
            .append(RcDoc::space())
            .append(RcDoc::text("in"))
            .append(RcDoc::line())
    }
}

impl PrettyPrint for Variable {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text(&self.0)
    }
}

impl PrettyPrint for Value {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Self::Variable(var) => var.to_doc(),
            Self::Op { op, vs, ds } => {
                let mut doc = op.to_doc();
                if !vs.is_empty() || !ds.is_empty() {
                    doc = doc.append(RcDoc::text("("));
                    if !vs.is_empty() {
                        doc = doc.append(RcDoc::intersperse(
                            vs.iter().map(PrettyPrint::to_doc),
                            RcDoc::text(",").append(RcDoc::space()),
                        ));
                    }
                    if !ds.is_empty() {
                        if !vs.is_empty() {
                            doc = doc.append(RcDoc::text(";")).append(RcDoc::space());
                        }
                        doc = doc.append(RcDoc::intersperse(
                            ds.iter().map(PrettyPrint::to_doc),
                            RcDoc::text(",").append(RcDoc::space()),
                        ));
                    }
                    doc = doc.append(RcDoc::text(")"));
                }
                doc
            }
        }
    }
}

impl PrettyPrint for Op {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Self::Plus => RcDoc::text("plus"),
            Self::Minus => RcDoc::text("minus"),
            Self::Times => RcDoc::text("times"),
            Self::Div => RcDoc::text("div"),
            Self::Rem => RcDoc::text("rem"),
            Self::And => RcDoc::text("and"),
            Self::Or => RcDoc::text("or"),
            Self::Not => RcDoc::text("not"),
            Self::Eq => RcDoc::text("eq"),
            Self::Neq => RcDoc::text("neq"),
            Self::Lt => RcDoc::text("lt"),
            Self::Leq => RcDoc::text("leq"),
            Self::Gt => RcDoc::text("gt"),
            Self::Geq => RcDoc::text("geq"),
            Self::App => RcDoc::text("app"),
            Self::Lambda => RcDoc::text("lambda"),
            Self::Unit => RcDoc::text("unit"),
            Self::Seq => RcDoc::text("seq"),
            Self::Atom => RcDoc::text("atom"),
            Self::Deref => RcDoc::text("deref"),
            Self::Assign => RcDoc::text("assign"),
            Self::Bool(b) => RcDoc::as_string(b),
            Self::Number(n) => RcDoc::as_string(n),
            Self::String(s) => RcDoc::text("\"")
                .append(RcDoc::as_string(s))
                .append(RcDoc::text("\"")),
            Self::Identifier(s) => RcDoc::as_string(s),
        }
    }
}

impl PrettyPrint for Thunk {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text("\\")
            .append(RcDoc::intersperse(
                self.args.iter().map(PrettyPrint::to_doc),
                RcDoc::space(),
            ))
            .append(RcDoc::space())
            .append(RcDoc::text("."))
            .append(if self.body.binds.is_empty() {
                RcDoc::space().append(self.body.to_doc())
            } else {
                RcDoc::line()
                    .append(self.body.to_doc())
                    .nest(4)
                    .append(RcDoc::line())
            })
    }
}

#[cfg(test)]
mod tests {
    use dir_test::{dir_test, Fixture};
    use insta::assert_snapshot;

    use crate::{language::spartan::Expr, prettyprinter::PrettyPrint};

    #[allow(clippy::needless_pass_by_value)]
    #[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "**/*.sd", loader: crate::language::spartan::tests::parse_sd, postfix: "pretty_print")]
    fn pretty_print(fixture: Fixture<(&str, Expr)>) {
        let (name, expr) = fixture.content();
        assert_snapshot!(format!("pretty_print_{name}"), expr.to_pretty());
    }
}
