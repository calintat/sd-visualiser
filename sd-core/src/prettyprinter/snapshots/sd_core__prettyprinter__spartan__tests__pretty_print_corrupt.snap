---
source: sd-core/src/prettyprinter/spartan.rs
expression: expr.to_pretty()
---
bind x = 1 in
bind y = plus(x, x . minus(x)) in
y
